### ISSUE TITLE
buildroot check now matching a script that was "undetected" for 10+ years

### ISSUE THREAD

**Issue by darix** [rpm-software-management/rpmlint#1350]
https://build.opensuse.org/projects/home:darix:apps/packages/ruby-common/files/gem_install.sh?expand=1

is that an intended real match or some false positive from the check?

**Comment by ana** [comment]
This comes from https://build.opensuse.org/request/show/1280116

it could be a regression from:
```
* Fixup fallback regexp for buildroot detection
```

**Comment by darix** [comment]
probably fallout from removing the anchoring in the fallback regexp.

**Comment by danigm** [comment]
The buildroot detection regex now match this script, when we update to support rpm 4.20 here:
https://github.com/rpm-software-management/rpmlint/commit/2a5b7a4de4d7c0723f5b76064e8adf6f7f2977f8

The actual regex can be found here:
https://github.com/rpm-software-management/rpmlint/blob/main/rpmlint/checks/BuildRootAndDateCheck.py#L20-L25

```
    def __init__(self, config, output):
        super().__init__(config, output, r'.*')
        self.looksliketime = re.compile('(2[0-3]|[01]?[0-9]):([0-5]?[0-9]):([0-5]?[0-9])')
        self.istoday = re.compile(time.strftime('%b %e %Y'))
        self.prepare_regex(rpm.expandMacro('%{?buildroot}') or '/.{1,40}/BUILDROOT/')

    def prepare_regex(self, buildroot):
        for m in ('name', 'version', 'release', 'NAME', 'VERSION', 'RELEASE'):
            buildroot = buildroot.replace('%%{%s}' % (m), r'[\w\!-\.]{1,20}')
        self.lookslikebuildroot = re.compile(buildroot)
```

**Comment by smithfarm** [comment]
It looks like @dirkmueller authored the change? Maybe he could weigh in.

**Comment by dirkmueller** [comment]
but with rpm 4.20 it is just `/BUILDROOT` iirc. to me it looks like a valid match.

**Comment by darix** [comment]
How about this? that should actually you know match what rpm 4.20 will use so we should only get files that match the current buildroot.

```
diff --git a/rpmlint/checks/BuildRootAndDateCheck.py b/rpmlint/checks/BuildRootAndDateCheck.py
index bf56fd06..486d3128 100644
--- a/rpmlint/checks/BuildRootAndDateCheck.py
+++ b/rpmlint/checks/BuildRootAndDateCheck.py
@@ -17,7 +17,14 @@ class BuildRootAndDateCheck(AbstractFilesCheck):
         super().__init__(config, output, r'.*')
         self.looksliketime = re.compile('(2[0-3]|[01]?[0-9]):([0-5]?[0-9]):([0-5]?[0-9])')
         self.istoday = re.compile(time.strftime('%b %e %Y'))
-        self.prepare_regex(rpm.expandMacro('%{?buildroot}') or '/.{1,40}/BUILDROOT/')
+        #
+        # in rpm 4.20 the expandMacro will return an empty string and so we run
+        # into the or case.
+        # return a string that is modelled after the actual path that rpm 4.20
+        # is using so we can actually match what the rpm 4.20 path will be and
+        # not just any random string in a file that references buildroot.
+        #
+        self.prepare_regex(rpm.expandMacro('%{?buildroot}') or '/.{1,40}/rpmbuild/%{name}-%{version}-build/BUILDROOT/')

     def prepare_regex(self, buildroot):
         for m in ('name', 'version', 'release', 'NAME', 'VERSION', 'RELEASE'):
```

**Comment by danigm** [comment]
> [@danigm](https://github.com/danigm) but this hardcodes a suseism in the check. there is no such thing in the upstream rpm versioning. it is simply appending /BUILDROOT/ to the builddir. I don't think that is a good change.

Okay, in that case, the correct way of checking this is what we have right now, and for the specific case of `ruby-common` package we just need a `rpmlintrc` file.

**Comment by dirkmueller** [comment]
which is what I said before ;-)

**Comment by dirkmueller** [comment]
we can do a middleground though, e.g. test for

`or '/%{name}-%{version}-build/BUILDROOT/'`

which happens to break the example in this issue, but matches the default (see https://github.com/rpm-software-management/rpm/blob/rpm-4.20.1-release/build/parsePreamble.c#L1290 )

**Comment by dirkmueller** [comment]
(technically we could use %{NAME} also but the regexp would be the same.
